### DISCUSSION TITLE
Using react-query or swr with Supabase?

### DISCUSSION THREAD

**Discussion by emschwartz** [supabase/supabase#471] [Questions]
Does anyone have experience using supabase with `react-query` or `swr`? Do you prefer one or the other? Why?

**Comment by kiwicopple** [comment]
At Supabase, we put our client behind a Next.js API function, then use `swr` (we didn't have auth when we first started). It will be interesting to see what others do as we will probably move to client -> database for a lot of data fetching. I've never used `react-query` but `swr` has been great.

**Comment by MildTomato** [comment]
We've been trying both at Supabase, in our hosted dashboard.
They both have advantages and disadvantages but as far as I know, they are unrelated to using it with supabase-js.

Perhaps @alaister has an opinion though, as he was doing POC in our dashboard using react query.

**Reply by drewbietron** [reply]
@MildTomato Wondering if you all have tried using React Query with the new v2 lib.  Would love to hear some insights into how you implemented it!  I [asked a question earlier about this](https://github.com/supabase/supabase/discussions/9530).

**Comment by psteinroe** [comment]
If you plan to use swr with supabase, be sure to check out [Supabase Cache Helpers](https://github.com/psteinroe/supabase-cache-helpers). It integrates supabase-js with swr and comes with mutation, subscription and convenience hooks for pagination as well as storage support. All logic is split up in reusable packages, hence adding support for `react-query` should be quite straight forward too.

```tsx
   // Define the query. Check out the other query hooks for pagination etc.
    const { data, count } = useQuery(
    client
        .from("contact")
        .select("id,username,ticket_number", { count: "exact" })
        .eq("username", "myusername"),
    "multiple",
    {
        revalidateOnFocus: false,
        revalidateOnReconnect: false,
    }
    );

    // Define the mutation. Update, Upsert, and Delete are also supported.
    const [insert] = useInsertMutation(client.from("contact"), "single");

    // Subscriptions are also supported.
    const { status } = useSubscription(
        client.channel('random'),
        {
            event: "*",
            table: "contact",
            schema: "public",
        },
        ["id"],
        { callback: (payload) => console.log(payload) }
    );

```
